proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+	We have chosen against this alternative, because we do not think any existing rewrite rules depend critically on the previous behaviour and we expect error messages and warnings can still be written for the most common mistakes with a bit more effort.
+
+3. 	Use lambda binders instead of applications to figure out the scope of local variables automatically.
+	For example the "mapList" rule could look like this:
+	::
+
+		"mapList" [1]  forall f.    foldr (\x ys -> f : ys) [] = map (\x -> f)
+
+	Where the the rule matcher would recognise that the ``\x ->`` binders on the left and the right is the same.
+	From this we could deduce that the variables ``x`` should be allowed to occur in ``f``.
+	We have not chosen this syntax because it is less explicit about which locally bound variables are allowed to occur in which template variables.
+
+4. 	Implement more powerful higher order matching, for example as proposed by De Moor and Sittampalam in `"Higher-order matching for program transformation" <https://www.sciencedirect.com/science/article/pii/S0304397500004023>`_.
+
+	They show an example of higher order matching that is not covered by this proposal, namely the template ``forall f x. f x``.
+	Here they apply one template variable ``f`` to another template variable ``x``.
+	This often leads to ambiguity.
+	For example if we match that template against the term 0 we can get many possible substitutions: ::
+
+		[f :-> \a -> a, x :-> 0]
+		[f :-> \a -> 0]
+		[f :-> \g -> g 0, x :-> \a -> a]
+		[f :-> \g -> g (g 0), x :-> \a -> a]
+
+	We expect that this alternative would require much more significant changes to the rule matcher in GHC.
+
+
+Unresolved Questions
+--------------------
+
+1. 	What to do with polymorphic template variables?
+	Consider the code:
+	::
+
+		foo :: (forall a. [a] -> Int) -> Int
+		foo len = len [1,2,3] + len "abc"
+		{-# NOINLINE foo #-}
+
+		{-# RULES "foo" forall (f :: forall a. [a] -> Int). foo (\xs -> 1 + f xs) = 2 + foo f #-}
+
+	Here, the template variable ``f`` has a polymorphic type.
+	With explicit type abstractions and applications the rule looks like this:
+	::
+
+		{-# RULES "foo" forall (f :: forall a. [a] -> Int). foo (/\a. \(xs::[a]) -> 1 + f @a xs) = 2 + foo f #-}
+
+	The proposal could be change such that this rule would match the expression:
+	::
+
+		foo (/\b. \(ys::[b]). 1 + (reverse @b (take @b 3 ys)))
+
+
+	However, if we change the type of the template variable ``f`` to ``forall a. a -> Int``, then the rule with explicit type abstractions and applications looks like this:
+	::
+
+		{-# RULES "foo" forall (f :: forall a. a -> Int). foo (/\a. \(xs::[a]) -> 1 + f @[a] xs) = 2 + foo f #-}
+
+	(Note: we assume deep subsumption here for simplicity of presentation)
+
+	Now ``@[a]`` is no longer a plain locally bound variable, so this is no longer a higher order pattern.
```

Comment:
If we implement sufficient extensions in the future, such as allowing invertible arguments in higher-order patterns, then we'll be able to allow this refactoring, because `[]` is injective (and thus invertible). E.g., we could push `\@a -> @[a]` into our subsitution and presume that `f` really abstracts over `@a`.

Perhaps it's best to wait until we have implemented the extension on the term level.